syntax = "proto3";

package customer;

service CustomerService {
  rpc CreateCustomer (CreateCustomerRequest) returns (CreateCustomerResponse);
  rpc GetCustomerById (GetCustomerByIdRequest) returns (GetCustomerByIdResponse);
  rpc UpdateCustomer (UpdateCustomerRequest) returns (UpdateCustomerResponse);
  rpc DeleteCustomer (DeleteCustomerRequest) returns (DeleteCustomerResponse);
}

message Address {
  string street = 1;
  string city = 2;
  string state = 3;
  string zip = 4;
  string country = 5;
}

message Customer {
  string customerid = 1;
  string firstname = 2;
  string lastname = 3;
  string email = 4;
  string phone = 5;
  Address address = 6;
}

message CreateCustomerRequest {
  Customer customer = 1;
}

message CreateCustomerResponse {
  bool ok = 1;
  string message = 2;
}

message GetCustomerByIdRequest {
  string customerid = 1;
}

message GetCustomerByIdResponse {
  bool ok = 1;
  string message = 2;
  Customer customer = 3;
}

message UpdateCustomerRequest {
  string customerid = 1;
  Customer customer = 2; // fields you set will be applied (we'll ignore empty strings)
}

message UpdateCustomerResponse {
  bool ok = 1;
  string message = 2;
}

message DeleteCustomerRequest {
  string customerid = 1;
}

message DeleteCustomerResponse {
  bool ok = 1;
  string message = 2;
}
